### ISSUE TITLE
DualView::view<T>() always returns the device view, if memspace is CudaUVMSpace

### ISSUE THREAD

**Issue by brian-kelley** [kokkos/kokkos#3850]
TLDR: ``DualView::view<T>`` should use the same type logic as ``sync<T>`` and ``modify<T>`` to decide whether to return device or host view, not just check if the requested device's memory space matches t_dev's memory space.

@crtrott @kyungjoo-kim I think this may explain some of the unexpected behavior you guys were talking about on Slack recently.

The behavior of DualView's sync/modify/view which are templated on device have inconsistent behavior.

When the DualView's device type is ``Device<Cuda, CudaUVMSpace>``, the ``t_host`` and ``t_dev`` have the same memory space and alias each other, but have different execution spaces (``Serial`` and ``Cuda``). That's what they should be. Plugging in the full devices ``Device<Serial, CudaUVMSpace>`` and ``Device<Cuda, CudaUVMSpace>`` into ``DualView::template sync`` and ``DualView::template modify`` has the expected behavior.

However, ``DualView::template view`` with these device types both return the "device" view, the Cuda/CudaUVMSpace. It only looks at the argument Device's memory space to decide which view to return (unlike sync<T> and modify<T>, which also look at execution space).

Our Tpetra wrapper for this (``MultiVector::template getLocalView<Device>``) uses the return type of this to determine whether to do ``sync_host``/``modify_host`` or ``sync_device``/``modify_device``. So we have effectively always been taking the device path, but this isn't correct when combined with other syncs and modifies. The modify flags end up wrong, so the sync is skipping the deep_copy, which we rely on to fence.

The important parts of this output are the types of the views returned by ``DualView::view<T>()``, 10 lines down or so.
```
Device view type: N6Kokkos4ViewIPdJNS_10LayoutLeftENS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEvEEE
      (its device_type): N6Kokkos6DeviceINS_4CudaENS_12CudaUVMSpaceEEE
Host view type: N6Kokkos4ViewIPdJNS_10LayoutLeftENS_6DeviceINS_6SerialENS_12CudaUVMSpaceEEEEEE
      (its device_type): N6Kokkos6DeviceINS_6SerialENS_12CudaUVMSpaceEEE


Marking host modified with templated version.
Need sync host/device: 0/1 <--- sync<T> works as expected
Clearing sync state.
Marking device modified with templated version.
Need sync host/device: 1/0
Clearing sync state.

Getting device view using templated version.
Its device_type: N6Kokkos6DeviceINS_4CudaENS_12CudaUVMSpaceEEE
Getting host view using templated version.

***** HERE: asked for Serial/CudaUVMSpace view, but got back the Cuda/CudaUVMSpace
Its device_type: N6Kokkos6DeviceINS_4CudaENS_12CudaUVMSpaceEEE

// (and then everything below this is as expected)
Marking host modified, then doing sync<Device<Cuda, CudaUVMSpace>>
Now, need sync host/device: 0/0
Marking host modified, then doing sync<Cuda>
Now, need sync host/device: 0/0
Marking device modified, then doing sync<Device<Serial, CudaUVMSpace>>
Now, need sync host/device: 0/0
Marking device modified, then doing sync<Serial>
Now, need sync host/device: 0/0
Marking host modified, then doing sync<CudaUVMSpace>
Now, need sync host/device: 0/1
Marking device modified, then doing sync<CudaUVMSpace>
Now, need sync host/device: 1/0
```

Reproducer:
```
#include <Kokkos_Core.hpp>
#include <Kokkos_DualView.hpp>

using ExecSpace = Kokkos::Cuda;
using MemSpace = Kokkos::CudaUVMSpace;
using DeviceType = Kokkos::Device<ExecSpace, MemSpace>;

using DualViewType = Kokkos::DualView<double*, Kokkos::LayoutLeft, DeviceType>;

int main()
{
  Kokkos::initialize();
  {
    DualViewType dv("myView", 100);
    dv.clear_sync_state();
    using d_device = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>;
    using h_device = Kokkos::Device<Kokkos::Serial, Kokkos::CudaUVMSpace>;
    {
      auto v_d = dv.view_device();
      std::cout << "Device view type: " << typeid(v_d).name() << '\n';
      std::cout << "      (its device_type): " << typeid(typename decltype(v_d)::device_type).name() << '\n';
    }
    {
      auto v_h = dv.view_host();
      std::cout << "Host view type: " << typeid(v_h).name() << '\n';
      std::cout << "      (its device_type): " << typeid(typename decltype(v_h)::device_type).name() << '\n';
    }
    std::cout << "\n";
    std::cout << "Marking host modified with templated version.\n";
    dv.template modify<h_device>();
    std::cout << "Need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
    std::cout << "Clearing sync state.\n";
    dv.clear_sync_state();
    std::cout << "Marking device modified with templated version.\n";
    dv.template modify<d_device>();
    std::cout << "Need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
    std::cout << "Clearing sync state.\n";
    dv.clear_sync_state();
    std::cout << "\n";
    std::cout << "Getting device view using templated version.\n";
    {
      auto v_d = dv.template view<d_device>();
      std::cout << "Its device_type: " << typeid(typename decltype(v_d)::device_type).name() << '\n';
    }
    std::cout << "Getting host view using templated version.\n";
    {
      auto v_h = dv.template view<h_device>();
      std::cout << "Its device_type: " << typeid(typename decltype(v_h)::device_type).name() << '\n';
    }
    std::cout << "\n";
    std::cout << "Marking host modified, then doing sync<Device<Cuda, CudaUVMSpace>>\n";
    {
      dv.modify_host();
      dv.template sync<d_device>();
      std::cout << "Now, need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
      dv.clear_sync_state();
    }
    std::cout << "Marking host modified, then doing sync<Cuda>\n";
    {
      dv.modify_host();
      dv.template sync<typename d_device::execution_space>();
      std::cout << "Now, need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
      dv.clear_sync_state();
    }
    std::cout << "Marking device modified, then doing sync<Device<Serial, CudaUVMSpace>>\n";
    {
      dv.modify_device();
      dv.template sync<h_device>();
      std::cout << "Now, need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
      dv.clear_sync_state();
    }
    std::cout << "Marking device modified, then doing sync<Serial>\n";
    {
      dv.modify_device();
      dv.template sync<typename h_device::execution_space>();
      std::cout << "Now, need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
      dv.clear_sync_state();
    }
    std::cout << "Marking host modified, then doing sync<CudaUVMSpace>\n";
    {
      dv.modify_host();
      dv.template sync<Kokkos::CudaUVMSpace>();
      std::cout << "Now, need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
      dv.clear_sync_state();
    }
    std::cout << "Marking device modified, then doing sync<CudaUVMSpace>\n";
    {
      dv.modify_device();
      dv.template sync<Kokkos::CudaUVMSpace>();
      std::cout << "Now, need sync host/device: " << dv.need_sync_host() << "/" << dv.need_sync_device() << '\n';
      dv.clear_sync_state();
    }
  }
  Kokkos::finalize();
  return 0;
}
```

**Comment by DavidPoliakoff** [comment]
This is really great work, Brian, we'll definitely discuss this. Thanks for tracking it down

**Comment by kyungjoo-kim** [comment]
@brian-kelley I think that you got the right point.It sounds like that the main problem is not about fence but wrong modify flags with uvm.

**Comment by brian-kelley** [comment]
@kyungjoo-kim Yes, exactly. I know for sure that a few of the Ifpack2 test failures (with no CUDA_LAUNCH_BLOCKING) were caused by this exact issue because changing instances of "getLocalView<T>" to "getLocalView[Host/Device]" fixed them. But the others are caused by us not having the modify flags correct for another reason. For example, the Tpetra_idot 4-rank test.
